docs: document controller runtime, resources and talosctl get (#3249)
talos-systems/talos: website/content/docs/v0.9/Learn More/controllers-resources.md

---
title: "Controllers and Resources"
weight: 8
---

<!-- markdownlint-disable MD038 -->

> Talos implements concepts of *resources* and *controllers* to implement internal operations of the

Reviewer: ```suggestion
Talos implements concepts of *resources* and *controllers* to facilitate internal operations of the
```

> operating system. Talos resources are controllers are very similar to Kubernetes resources and controllers,

Reviewer: ```suggestion
operating system. Talos resources and controllers are very similar to Kubernetes resources and controllers,
```

> but there are some differences. Content of this document is not required to operate Talos, but it is

Reviewer: ```suggestion
but there are some differences. The content of this document is not required to operate Talos, but it is
```

> useful for troubleshooting Talos.

Author: I wonder if I should just drop `Talos` there, it should be clear without it.

> useful for troubleshooting Talos.

Reviewer 2: Should be clear without it.

Starting with Talos 0.9, most of the Kubernetes control plane boostrapping and operations is implemented
via controllers and resources which allows Talos to be reactive to configuration changes, environment
changes (e.g. time sync).

## Resources

> Resource captures piece of system state. Each resource belongs to some type which defines resource contents.

Reviewer: ```suggestion
A resource captures a piece of system state. Each resource belongs to a "Type" which defines resource contents.
```

Resource state can be split in two parts:

* metadata: fixed set of fields describing resource - namespace, type, ID, etc.
* spec: contents of the resource (depends on resource type).

Resource is uniquely identified by (`namespace`, `type`, `id`). Namespaces provide a way to avoid conflicts on
duplicate resource IDs.

At the moment of this writing, all resources are local to the node and stored in memory. So on every reboot
resource state is rebuilt from scratch (the only exception is `MachineConfig` resource which reflects current
machine config).

## Controllers

> Controllers are running as independent lightweight threads in Talos. Goal of the controller is to reconcile
> the state based on the inputs and update outputs.

Reviewer: ```suggestion
Controllers run as independent lightweight threads in Talos. The goal of the controller is to reconcile
the state based on inputs and eventually update outputs.
```

Controller can have any number of resource types (and namespaces) as input, in other words it watches
specified resources for changes and reconciles on input resource change. Controller might have additional
inputs: running reconcile on schedule, watching `etcd` keys, etc.

> Controller has single output: set of resources of fixed type in a fixed namespace. Only one controller
> can manage resource type in the namespace, so conflicts are avoided.

Reviewer: ```suggestion
A controller has a single output: a set of resources of fixed type in a fixed namespace. Only one controller
can manage resource type in the namespace, so conflicts are avoided.
```

172.20.0.2   meta        Namespace   secrets        1
```

Most of the time namespace flag (`--namespace`) can be omitted, as `ResourceDefinition` contains default
namespace which is used if no namespace is given:

```bash
$ talosctl get resourcedefinitions resourcedefinitions.meta.cosi.dev -o yaml
node: 172.20.0.2
metadata:
    namespace: meta
    type: ResourceDefinitions.meta.cosi.dev
    id: resourcedefinitions.meta.cosi.dev
    version: 1
    phase: running
spec:
    type: ResourceDefinitions.meta.cosi.dev
    displayType: ResourceDefinition
    aliases:
        - resourcedefinitions
        - resourcedefinition
        - resourcedefinitions.meta
        - resourcedefinitions.meta.cosi
        - rd
        - rds
    printColumns: []
    defaultNamespace: meta
```

Resource definition also contains type aliases which can be used interchangeably with canonical resource name:

```bash
$ talosctl get ns config
NODE         NAMESPACE   TYPE        ID             VERSION
172.20.0.2   meta        Namespace   config         1
```

### Output

Command `talosctl get` supports following output modes:

* `table` (default) prints resource list as a table

> * `yaml` prints pretty formatted resources with details, including full metadata spec, this format carries most

Reviewer: ```suggestion
* `yaml` prints pretty formatted resources with details, including full metadata spec. This format carries most
```

  details from the backend resource (e.g. comments in `MachineConfig` resource)
* `json` prints same information as `yaml`, some additional details (e.g. comments) might be lost,
  this format is useful for automated processing with tools like `jq`

### Watching Changes

> If flag `--watch` is appended to the `talosctl get` command, command switches to watch mode. If list of resources

Reviewer: ```suggestion
If flag `--watch` is appended to the `talosctl get` command, the command switches to watch mode. If list of resources
```

172.20.0.2   +   runtime   Service   udevd   2   true   true
172.20.0.2   -   runtime   Service   timed   2   true   true
172.20.0.2   +   runtime   Service   timed   1   true   false
172.20.0.2       runtime   Service   timed   2   true   true
```

Column `*` specifies event type:

* `+` is created
* `-` is deleted
* ` ` is updated

In YAML/JSON output, field `event` is added to the resource representation to describe the event type.

### Examples

Getting machine config:

```bash
$ talosctl get machineconfig -o yaml
node: 172.20.0.2
metadata:
    namespace: config
    type: MachineConfigs.config.talos.dev
    id: v1alpha1
    version: 2
    phase: running
spec:
    version: v1alpha1 # Indicates the schema used to decode the contents.
    debug: false # Enable verbose logging to the console.
    persist: true # Indicates whether to pull the machine config upon every boot.
    # Provides machine specific configuration options.
...
```

Getting control plane static pod statuses:

```bash
$ talosctl get staticpodstatus
NODE         NAMESPACE      TYPE              ID                                                           VERSION   READY

> 172.20.0.2   controlplane   StaticPodStatus   kube-system/kube-apiserver-talos-default-master-1            3         True

Author: example here should be accurate, full type name is `StaticPodStatuses.kubernetes.talos.dev`, but we show `DisplayType` which is `StaticPodStatus` (similar to Kubernetes, which doesn't use full type name in table output).

> 172.20.0.2   controlplane   StaticPodStatus   kube-system/kube-apiserver-talos-default-master-1            3         True

Reviewer 3: yeah, sorry. I just had an old version of talsoctl :facepalm:
